lib/AST/TypeJoinMeet.cpp

```diff
@@ -17,100 +17,197 @@
+
+    return TypeJoin(second).visit(first);
+  }
+};
+
+Type TypeJoin::getSuperclassJoin(Type first, Type second) {
+  if (!first || !second)
+    return TypeJoin::join(first, second);
+
+  // FIXME: Return Any
+  if (!first->mayHaveSuperclass() || !second->mayHaveSuperclass())
     return nullptr;
+
+  /// Walk the superclasses of `first` looking for `second`. Record them
+  /// for our second step.
+  llvm::SmallPtrSet<CanType, 8> superclassesOfFirst;
+  CanType canSecond = second->getCanonicalType();
+  for (Type super = first; super; super = super->getSuperclass()) {
+    CanType canSuper = super->getCanonicalType();
+
+    // If we have found the second type, we're done.
+    if (canSuper == canSecond) return super;
+
+    superclassesOfFirst.insert(canSuper);
   }
 
-  // If one or both of the types are optional types, look at the underlying
-  // object type.
+  // Look through the superclasses of second to determine if any were also
+  // superclasses of first.
+  for (Type super = second; super; super = super->getSuperclass()) {
+    CanType canSuper = super->getCanonicalType();
+
+    // If we found the first type, we're done.
+    if (superclassesOfFirst.count(canSuper)) return super;
+  }
+
+  // FIXME: Return Any
+  // There is no common superclass; we're done.
+  return nullptr;
+}
+
+Type TypeJoin::visitClassType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitBoundGenericClassType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitArchetypeType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitDynamicSelfType(Type second) {
+  return getSuperclassJoin(First, second);
+}
+
+Type TypeJoin::visitMetatypeType(Type second) {
+  assert(!First->mayHaveSuperclass() && !second->mayHaveSuperclass());
+
+  // FIXME: Return Any
+  if (First->getKind() != second->getKind())
+    return nullptr;
+
+  auto firstInstance = First->castTo<AnyMetatypeType>()->getInstanceType();
+  auto secondInstance = second->castTo<AnyMetatypeType>()->getInstanceType();
+
+  auto joinInstance = join(firstInstance, secondInstance);
+
+  // FIXME: Return Any
+  if (!joinInstance)
+    return nullptr;
+
+  return MetatypeType::get(joinInstance);
+}
+
+Type TypeJoin::visitExistentialMetatypeType(Type second) {
+  assert(!First->mayHaveSuperclass() && !second->mayHaveSuperclass());
+
+  // FIXME: Return Any
+  if (First->getKind() != second->getKind())
+    return nullptr;
+
+  auto firstInstance = First->castTo<AnyMetatypeType>()->getInstanceType();
+  auto secondInstance = second->castTo<AnyMetatypeType>()->getInstanceType();
+
+  auto joinInstance = join(firstInstance, secondInstance);
+
+  // FIXME: Return Any
+  if (!joinInstance)
+    return nullptr;
+
+  return ExistentialMetatypeType::get(joinInstance);
+}
+
+Type TypeJoin::visitBoundGenericEnumType(Type second) {
+  // FIXME: Return Any
+  if (First->getKind() != second->getKind())
```

Comment:
I'd boost this code up to `visitBoundGenericType`, so it can apply to all such types. Need to look at the parents, too.